### PR TITLE
pathlib for tests

### PR DESCRIPTION
**Pull request by cakester**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

#### PR contents

- [x] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [x] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [x] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description

Phased refactor adoption of loader/segmentation_pair.py as detailed in #872 using pathlib for using pathlib for using pathlib for tests folder instead of using os.path

## Linked issues

Step wise remediation of #872

### CODE REVIEW

File: testing/unit_tests/test_loader.py [hunk 1]
```diff
@@ -172,14 +172,14 @@ def test_load_dataset_2d_png(download_data_testing_test_files,
                                                              'dataset_type': 'training'}})
     fname_png = bids_df.df[bids_df.df['filename'] == data_lst[0]]['path'].values[0]
     fname_nii = imed_loader_utils.update_filename_to_nifti(fname_png)
-    assert os.path.exists(fname_nii) == 1
+    assert Path(fname_nii).exists() == 1
```
Comment (dyt811) [inline testing/unit_tests/test_loader.py:175]: Probably `is true`  or `==True` versus `==1`

File: testing/unit_tests/test_onnx.py [hunk 2]
```diff
@@ -16,29 +16,29 @@ def setup_function():
...
+PATH_MODEL_PT = PATH_MODEL_ONNX.with_suffix('.pt')
 LENGTH_3D = (112, 112, 112)
 
 
 def test_onnx(download_data_testing_test_files):
     model = imed_models.Modified3DUNet(1, 1)
-    if not os.path.exists(PATH_MODEL):
-        os.mkdir(PATH_MODEL)
+    if not PATH_MODEL.exists():
+        PATH_MODEL.mkdir()
```
Comment (dyt811) [inline testing/unit_tests/test_onnx.py:29]: ditto on earlier `mkdir()` arguments

File: testing/unit_tests/test_postprocessing.py [hunk 3]
```diff
@@ -173,7 +173,7 @@ def test_label_file_from_coordinates(download_data_testing_test_files):
     coord = [[0, 0, 0]]
     # load test image
     nifti = nib.load(
-        os.path.join(__data_testing_dir__, 'sub-unf01/anat/sub-unf01_T1w.nii.gz'))
+        Path(__data_testing_dir__, 'sub-unf01/anat/sub-unf01_T1w.nii.gz'))
```
Comment (dyt811) [inline testing/unit_tests/test_postprocessing.py:176]: Good call on `_stringify_path` can convert `Path` object for `nib.load`

File: testing/unit_tests/test_split_dataset.py [hunk 4]
```diff
@@ -7,9 +6,10 @@
 from ivadomed.loader import utils as imed_loader_utils
 from testing.unit_tests.t_utils import create_tmp_dir, __tmp_dir__
 from testing.common_testing_util import remove_tmp_dir
+from pathlib import Path
 
-PATH_DATA = os.path.join(__tmp_dir__, 'bids')
-LOG_PATH = os.path.join(__tmp_dir__, 'log')
+PATH_DATA = Path(__tmp_dir__, 'bids')
+LOG_PATH = Path(__tmp_dir__, 'log')
```
Comment (dyt811) [inline testing/unit_tests/test_split_dataset.py:12]: Consistency here for variable naming

File: testing/unit_tests/test_split_dataset.py [hunk 5]
```diff
@@ -23,12 +23,12 @@ def load_dataset(split_params):
     create_jsonfile()
 
     # Create log path
-    if not os.path.isdir(LOG_PATH):
-        os.mkdir(LOG_PATH)
+    if not LOG_PATH.is_dir():
+        LOG_PATH.mkdir()
```
Comment (dyt811) [inline testing/unit_tests/test_split_dataset.py:27]: Same about arguments here

File: testing/unit_tests/test_split_dataset.py [hunk 6]
```diff
@@ -162,8 +162,8 @@ def test_per_center_balance(split_params):
 
 def create_tsvfile():
     # Create data path
-    if not os.path.isdir(PATH_DATA):
-        os.mkdir(PATH_DATA)
+    if not PATH_DATA.is_dir():
+        PATH_DATA.mkdir()
```
Comment (dyt811) [inline testing/unit_tests/test_split_dataset.py:166]: The optional arguments

File: testing/unit_tests/test_tensorboard_save.py [hunk 7]
```diff
@@ -19,13 +19,13 @@ def test_tensorboard_save():
     inp = torch.tensor(np.zeros((1, 1, 15, 15)))
     gt = torch.tensor(np.zeros((1, 1, 15, 15)))
     pred = torch.tensor(np.zeros((1, 1, 15, 15)))
-    dpath = os.path.join(__tmp_dir__, "test_tensorboard_save")
-    os.makedirs(dpath)
-    writer = SummaryWriter(log_dir=dpath)
+    dpath = Path(__tmp_dir__, "test_tensorboard_save")
+    dpath.mkdir(parents=True)
```
Comment (dyt811) [inline testing/unit_tests/test_tensorboard_save.py:23]: I recommend adding optional ignore error argument.

File: testing/unit_tests/test_testing.py [hunk 8]
```diff
@@ -108,14 +108,14 @@ def test_inference(download_data_testing_test_files, transforms_dict, test_lst,
 
     metric_mgr = imed_metrics.MetricManager(metric_fns)
 
-    if not os.path.isdir(__output_dir__):
-        os.makedirs(__output_dir__)
+    if not __output_dir__.is_dir():
+        __output_dir__.mkdir(parents=True)
```
Comment (dyt811) [inline testing/unit_tests/test_testing.py:112]: Let's just make all mkdir consistent with regard to optional parameters.

File: testing/unit_tests/test_testing.py [hunk 9]
```diff
@@ -200,18 +200,18 @@ def test_inference_2d_microscopy(download_data_testing_test_files, transforms_di
         model.cuda()
     model.eval()
 
-    if not os.path.isdir(__output_dir__):
-        os.makedirs(__output_dir__)
+    if not __output_dir__.is_dir():
+        __output_dir__.mkdir(parents=True)
```
Comment (dyt811) [inline testing/unit_tests/test_testing.py:204]: Same as above.

File: testing/unit_tests/test_testing.py [hunk 10]
```diff
@@ -200,18 +200,18 @@ def test_inference_2d_microscopy(download_data_testing_test_files, transforms_di
...
                                                    model_params=model_params,
                                                    testing_params=testing_params,
-                                                   ofolder=__output_dir__,
+                                                   ofolder=str(__output_dir__),
                                                    cuda_available=cuda_available)
 
-    assert len([x for x in os.listdir(__output_dir__) if x.endswith(".nii.gz")]) == len(test_lst)
-    assert len([x for x in os.listdir(__output_dir__) if x.endswith(".png")]) == 2*len(test_lst)
+    assert len([x for x in __output_dir__.iterdir() if x.name.endswith(".nii.gz")]) == len(test_lst)
+    assert len([x for x in __output_dir__.iterdir() if x.name.endswith(".png")]) == 2*len(test_lst)
```
Comment (dyt811) [inline testing/unit_tests/test_testing.py:214]: I think these can just use `Path().suffix` ?

**Review by dyt811 — CHANGES_REQUESTED** [review]
Minor changes needed for  a few `with Path()` handling to ensure they are opened properly with Path object.

File: testing/unit_tests/test_testing.py (hunk 10 above)
Comment (cakester) [reply testing/unit_tests/test_testing.py:214]: >>> from pathlib import Path
>>> a=Path("a.nii.gz")
>>> a.suffix
'.gz'

Comment (dyt811) [reply testing/unit_tests/test_testing.py:215]: I think we can potentially use [suffixes](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.suffixes)

Comment (cakester) [reply testing/unit_tests/test_testing.py:215]: I think checking a string with ends with is easier than checking a list

**Review by dyt811 — APPROVED** [review]
Thanks to Hanna's hard work, we should have a much more consistent Pathlib integration across the entire code base for long term cross platform compatibility.
